Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -32,311 +32,259 @@
+        return false;
+    }
 
-        // This was a simple content separator not a closing one.
-        // Go to before the content processing.
-        m_state = State::InBoundary;
-        break;
+    case State::Terminating: {
+        m_client->didCompleteFromMultipart();
+        m_state = State::End;
+        return false;
     }
+
     case State::End:
-        // We are done. Nothing to do anymore.
-        return false;
-    default:
-        ASSERT_NOT_REACHED();
         return false;
     }
 
-    return true; // There are still things to process, so go for it.
+    return false;
 }
 
-bool CurlMultipartHandle::checkForBoundary(size_t& boundaryStartPosition, size_t& lastPartialMatchPosition)
+bool CurlMultipartHandle::shouldTerminateIfNeeded()
 {
+    if (m_didCompleteMessage) {
+        m_state = State::Terminating;
+        return true;
+    }
+
+    return false;
+}
+
+CurlMultipartHandle::FindBoundaryResult CurlMultipartHandle::findBoundary()
+{
+    FindBoundaryResult result;
+
     auto contentLength = m_buffer.size();
+    const auto contentStartPtr = m_buffer.data();
+    const auto contentEndPtr = contentStartPtr + contentLength;
+
+    auto boundaryLength = m_boundary.length();
+    auto boundaryStartPtr = m_boundary.data();
+
+    auto matchedStartPtr = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, boundaryStartPtr, boundaryLength));
+    if (!matchedStartPtr) {
+        if (!m_didCompleteMessage) {
+            // Not enough data to check the boundary (Temporarily retain "Initial CRLF + (boundary - 1)" bytes for the next search.)
+            result.dataEnd = std::max(int(contentLength) - int(boundaryLength) + 1 - 2, 0);
+        } else
+            result.dataEnd = contentLength;
+
+        result.processed = result.dataEnd;
+        return result;
+    }
 
-    boundaryStartPosition = notFound;
-    lastPartialMatchPosition = contentLength;
+    auto matchedEndPtr = matchedStartPtr + boundaryLength;
 
-    if (contentLength < m_boundary.length())
-        return false;
+    // The initial CRLF is considered to be attached to the boundary delimiter line rather than
+    // part of the preceding part. (See RFC2046 [5.1.1. Common Syntax])
+    if (matchedStartPtr - contentStartPtr >= 2 && !memcmp(matchedStartPtr - 2, "\r\n", 2))
+        matchedStartPtr -= 2;
+    else if (matchedStartPtr - contentStartPtr >= 1 && !memcmp(matchedStartPtr - 1, "\n", 1))
+        matchedStartPtr--;
 
-    const auto content = m_buffer.data();
+    result.dataEnd = matchedStartPtr - contentStartPtr;
+    result.processed = result.dataEnd;
 
-    for (size_t i = 0; i < contentLength - m_boundary.length(); ++i) {
-        auto length = matchedLength(content + i);
-        if (length == m_boundary.length()) {
-            boundaryStartPosition = i;
-            return true;
+    // Check the Close Delimiter
+    if (contentEndPtr - matchedEndPtr < 2) {
+        if (!m_didCompleteMessage) {
```

Comment:
I think this code ends in an infinite loop if m_didCompleteMessage here. You should have a test case.

Reply:
I don't think so. If m_didCompleteMessage is true, one of the following three cases will handle.

https://github.com/WebKit/WebKit/pull/17403/files#diff-9748a6c95db83f4a038a66c392226087d69df3953308a3f96c081618d378e750R250-R258
```
    else if (contentEndPtr - matchedEndPtr >= 1 && !memcmp(matchedEndPtr, "\n", 1))
        matchedEndPtr++;
    else if (matchedEndPtr >= contentEndPtr) {
        // Not enough data to check the boundary
        return result;
    } else {
        result.isSyntaxError = true;
        return result;
    }
```

After that, it ends with an error or the shouldTerminateIfNeeded() condition is met, so it transitions to State::Terminating and ends. Therefore, it will not enter an infinite loop.

https://github.com/WebKit/WebKit/pull/17403/files#diff-9748a6c95db83f4a038a66c392226087d69df3953308a3f96c081618d378e750R135-R136
```
            if (shouldTerminateIfNeeded())
                return true;
```

https://github.com/WebKit/WebKit/pull/17403/files#diff-9748a6c95db83f4a038a66c392226087d69df3953308a3f96c081618d378e750R151-R152
```
            if (shouldTerminateIfNeeded())
                return true;
```

Reply:
You are right.